shell_integration/windows/OCUtil/CommunicationSocket.cpp

```diff
@@ -122,16 +121,21 @@ bool CommunicationSocket::ReadLine(wstring* response)
         DWORD totalBytesAvailable = 0;
 
         if (!PeekNamedPipe(_pipe, NULL, 0, 0, &totalBytesAvailable, 0)) {
-            Close();
             return false;
         }
         if (totalBytesAvailable == 0) {
-            return false;
+            return true;
         }
 
-        if (!ReadFile(_pipe, resp_utf8.data(), DWORD(resp_utf8.size()), &numBytesRead, NULL)) {
-            Close();
-            return false;
+        bool ok = ReadFile(_pipe, resp_utf8.data(), DWORD(resp_utf8.size()), &numBytesRead, &_overlapped);
+        if (!ok) {
+            if (GetLastError() == ERROR_IO_PENDING) {
+                WaitForSingleObject(_overlapped.hEvent, timeoutC);
+                ok = GetOverlappedResult(_pipe, &_overlapped, &numBytesRead, FALSE);
+            }
+            if (!ok) {
```

Comment:
Please add a comment here to explain this flow. The double-check on `ok` can be confusing